utility/unit_of_work/module.go

```diff
@@ -0,0 +1,209 @@
+	if err != nil {
+		u.logger.Fatal().Err(err).Msg("Updating the app hash failed. TODO: Look into roll-backing the entire commit...")
+	}
+	u.logger.Info().Str("state_hash", stateHash).Msgf("CreateAndApplyProposalBlock finished successfully")
+
+	return stateHash, txs, err
+}
+
+// CLEANUP: code re-use ApplyBlock() for CreateAndApplyBlock()
+func (u *baseUtilityUnitOfWork) ApplyBlock() (stateHash string, txs [][]byte, err error) {
+	lastByzantineValidators, err := u.prevBlockByzantineValidators()
+	if err != nil {
+		return "", nil, err
+	}
+
+	// begin block lifecycle phase
+	if err := u.beginBlock(lastByzantineValidators); err != nil {
+		return "", nil, err
+	}
+
+	mempool := u.GetBus().GetUtilityModule().GetMempool()
+
+	// deliver txs lifecycle phase
+	for index, txProtoBytes := range u.proposalBlockTxs {
+		tx, err := coreTypes.TxFromBytes(txProtoBytes)
+		if err != nil {
+			return "", nil, err
+		}
+		if err := tx.ValidateBasic(); err != nil {
+			return "", nil, err
+		}
+		// TODO(#346): Currently, the pattern is allowing nil err with an error transaction...
+		//             Should we terminate applyBlock immediately if there's an invalid transaction?
+		//             Or wait until the entire lifecycle is over to evaluate an 'invalid' block
+
+		// Validate and apply the transaction to the Postgres database
+		txResult, err := u.hydrateTxResult(tx, index)
+		if err != nil {
+			return "", nil, err
+		}
+
+		txHash, err := tx.Hash()
+		if err != nil {
+			return "", nil, err
+		}
+
+		// TODO: Need to properly add transactions back on rollbacks
+		if mempool.Contains(txHash) {
+			if err := mempool.RemoveTx(txProtoBytes); err != nil {
+				return "", nil, err
+			}
+			u.logger.Info().Str("tx_hash", txHash).Msg("Applying tx that WAS in the local mempool")
+		} else {
+			u.logger.Info().Str("tx_hash", txHash).Msg("Applying tx that WAS NOT in the local mempool")
+		}
+
+		if err := u.persistenceRWContext.IndexTransaction(txResult); err != nil {
+			u.logger.Fatal().Err(err).Msgf("TODO(#327): We can apply the transaction but not index it. Crash the process for now: %v\n", err)
+		}
+	}
+
+	// end block lifecycle phase
+	if err := u.endBlock(u.proposalProposerAddr); err != nil {
+		return "", nil, err
+	}
+	// TODO: @deblasis - this should be from a ReadContext (the ephemeral/staging one)
+	// return the app hash (consensus module will get the validator set directly)
+	stateHash, err = u.persistenceRWContext.ComputeStateHash()
+	if err != nil {
+		u.logger.Fatal().Err(err).Msg("Updating the app hash failed. TODO: Look into roll-backing the entire commit...")
+		return "", nil, utilTypes.ErrAppHash(err)
+	}
+	u.logger.Info().Msgf("ApplyBlock - computed state hash: %s", stateHash)
+
+	// return the app hash; consensus module will get the validator set directly
+	return stateHash, nil, nil
+}
+
+func (uow *baseUtilityUnitOfWork) Commit(quorumCert []byte) error {
+	// TODO: @deblasis - change tracking here
```

Comment:
What do you mean by tracking?

Reply:
As in: "keeping track of the changes"

The comments with my names are notes to self. In there I mean: apply the changes we tracked that we have to commit